newlib/libc/machine/riscv/memcpy.S

```diff
@@ -0,0 +1,295 @@
+/* Copyright (c) 2019  SiFive Inc. All rights reserved.
+
+   This copyrighted material is made available to anyone wishing to use,
+   modify, copy, or redistribute it subject to the terms and conditions
+   of the FreeBSD License.   This program is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY expressed or implied,
+   including the implied warranties of MERCHANTABILITY or FITNESS FOR
+   A PARTICULAR PURPOSE.  A copy of this license is available at
+   http://www.opensource.org/licenses.
+*/
+
+#include <sys/asm.h>
+
+.text
+.global memcpy
+.type	memcpy, @function
+memcpy:
+#if defined(PREFER_SIZE_OVER_SPEED) || defined(__OPTIMIZE_SIZE__)
+  mv a3, a0
+  beqz a2, 2f
+
+1:
+  lbu a4, 0(a1)
+  sb a4, 0(a3)
+  add   a2, a2, -1
+  add   a3, a3, 1
+  add   a1, a1, 1
+  bnez a2, 1b
+
+2:
+  ret
+
+#else
+    mv a3, a0
+    li    a4, SZREG
+    addi  sp, sp, -2 * SZREG
+#if __riscv_xlen == 32
+    sw    s0, 0(sp)
+    sw    s1, 4(sp)
+#else
+    sd    s0, 0(sp)
+    sd    s1, 8(sp)
+#endif
+    bleu  a2, a4, .Lcopy1
+
+
+#if defined(__riscv_misaligned_avoid) || defined(__riscv_misaligned_slow)
+    andi a4, a1, SZREG - 1
+    andi a5, a3, SZREG - 1
+    sub s0, a4, a5
+    beqz s0, .Lalign
+    li s0, SZREG
+    sub s0, s0, a5
+
+;# In the case where source and destination are not aligned, we align the
+;# destination, load SZREG bytes, one at a time, from source, and then store them
+;# to destination in one instruction.
+.Lalign_dest:
+    lbu a4, 0(a1)
+    sb  a4, 0(a3)
+    addi a1, a1, 1
+    addi a3, a3, 1
+    addi s0, s0, -1
+    addi a2, a2, -1
+
+    bnez s0, .Lalign_dest
+
+.Ldest_aligned:
+    li s0, SZREG
+    bleu a2, s0, .Lcopy1
+
+    lbu a4, 0(a1)
+    lbu a5, 1(a1)
+    lbu s0, 2(a1)
+    lbu s1, 3(a1)
+
+#if __riscv_xlen == 32
+
+#ifdef __riscv_zbkb
```

Comment:
@thebigclub : If multiple of you use this sort of thing, we should get this into a shared macro. 
Assuming this is the case, I would suggest that we deal with this when submitting a second patch that does this.